### PR TITLE
Task/shank 109 anonymizer

### PR DESCRIPTION
**Pull request by ilaflamme-va**
https://vasdvp.atlassian.net/browse/SHANK-109

### CODE REVIEW

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/FileBasedNames.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import lombok.NoArgsConstructor;
+import lombok.SneakyThrows;
+
+@NoArgsConstructor
+public class FileBasedNames implements Names {
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/FileBasedNames.java]: You should try this:
```
var f1 = new FileBasedNames();
var f2 = new FileBasedNames(List.of("mmm"));
println(f1.get(1));
println(f1.get(2));
```

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  @Override
+  public String synthesizeDate(String rawDate) {
+    if (StringUtils.isBlank(rawDate)) {
+      return null;
+    }
+    LocalDate date;
+    try {
+      date = LocalDate.parse(rawDate);
+      date = date.withMonth(date.getMonthValue() - 1);
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java]: date's should truncate to the _year_ I believe? @eclendenning-va

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Override
+  public String synthesizeDateTime(String rawDateTime) {
+    if (StringUtils.isBlank(rawDateTime)) {
+      return null;
+    }
+    LocalDateTime date;
+    try {
+      date = LocalDateTime.parse(rawDateTime);
+      date = date.withMonth(date.getMonthValue() - 1);
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java]: truncate to year

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/SyntheticData.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package gov.va.api.health.bulkfhir.anonymizer;
+
+import gov.va.api.health.dstu2.api.datatypes.HumanName;
+import java.util.List;
+
+public interface SyntheticData {
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/SyntheticData.java]: maybe a little doc as to _what_ and _why_ someone would want synthetic data

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/patient/AnonymizerPatientSamples.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
+package gov.va.api.health.bulkfhir.anonymizer.patient;
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/patient/AnonymizerPatientSamples.java]: I don't think samples should be production code, move to `src/test/java`

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java (hunk 2 above)
Comment (eclendenning-va) [reply anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java]: "All elements of dates (except year) for dates directly related to an individual, including birth date, admission date, discharge date, date of death; and all ages over 89 and all elements of dates (including year) indicative of such age, except that such ages and elements may be aggregated into a single category of age 90 or older."

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/patient/PatientAnonymizer.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        // remove identifier
+        .implicitRules(resource.implicitRules())
+        .language(resource.language())
+        .link(resource.link())
+        .managingOrganization(resource.managingOrganization())
+        .maritalStatus(resource.maritalStatus())
+        .meta(resource.meta())
+        .modifierExtension(resource.modifierExtension())
+        .multipleBirthBoolean(resource.multipleBirthBoolean())
+        // Remove multipleBirthInteger
```
Comment (bschofield-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/patient/PatientAnonymizer.java]: Shouldn't there be some logic here if `multipleBirthInteger` is supplied? Choice structures are _one of_.  So something like this maybe:

```
Boolean sanitizedMultipleBirth = resource.multipleBirthBoolean();
Integer mi = resource.multipleBirthInteger();
if (mi != null) { sanitizedMultipleBirth = (mi == 0?false:true); }
```

File: anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  @Override
+  public String synthesizeDateTime(String rawDateTime) {
+    if (StringUtils.isBlank(rawDateTime)) {
+      return null;
+    }
+    LocalDateTime date;
+    try {
+      date = LocalDateTime.parse(rawDateTime);
```
Comment (eclendenning-va) [inline anonymizer/src/main/java/gov/va/api/health/bulkfhir/anonymizer/ResourceBasedSyntheticData.java]: The time would get through the anonymization , and spelling of default in the log

File: anonymizer/src/test/java/gov/va/api/health/bulkfhir/anonymizer/ClassPathResourceBasedNamesTest.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.junit.jupiter.api.Test;
+
+@Slf4j
+public class ClassPathResourceBasedNamesTest {
+
+  private List<String> testNames = List.of("Washington", "Adams", "Jefferson", "Monroe", "Madison");
+
+  @Test
+  void getName() {
```
Comment (bschofield-va) [inline anonymizer/src/test/java/gov/va/api/health/bulkfhir/anonymizer/ClassPathResourceBasedNamesTest.java]: Change not required, but something to think about in the future... name tests with their expectation, e.g. `getNameMapsNumbersRegardlessOfListSize()` or something like that.

Comment (ilaflamme-va) [reply anonymizer/src/test/java/gov/va/api/health/bulkfhir/anonymizer/ClassPathResourceBasedNamesTest.java]: :+1:

**Pull request merged**